src/D2L.CodeStyle.Analyzers/Language/AttributeAliasesAnalyzer.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class AttributeAliasesAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.AliasingAttributeNamesNotSupported
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		private void CompilationStart( CompilationStartAnalysisContext context ) {
+
+			context.RegisterSyntaxNodeAction(
+					c => AnalyzeAttribute( c, (AttributeSyntax)c.Node ),
+					SyntaxKind.Attribute
+				);
+		}
+
+		private void AnalyzeAttribute(
+				SyntaxNodeAnalysisContext context,
+				AttributeSyntax attribute
+			) {
+
+			// if it's not an identifier, then it's qualified and not aliasing the attribute name
+			if( attribute.Name is not IdentifierNameSyntax attributeName  ) {
+				return;
+			}
+
+			IEnumerable<IdentifierNameSyntax> usingAliases = GetUsingAliases( attribute );
+			foreach( IdentifierNameSyntax usingAlias in usingAliases ) {
+
+				if( !IsEquivalentToUsingAlias( attributeName, usingAlias ) ) {
+					continue;
+				}
+
+				Diagnostic d = Diagnostic.Create(
+						Diagnostics.AliasingAttributeNamesNotSupported,
+						attribute.GetLocation()
+					);
+
+				context.ReportDiagnostic( d );
+				return;
+			}
+		}
+
+		private static bool IsEquivalentToUsingAlias(
+				IdentifierNameSyntax attributeIdentifier,
+				IdentifierNameSyntax usingAliasIdentifier
+			) {
+
+			string attributeName = attributeIdentifier.ToString();
+			string usingAlias = usingAliasIdentifier.ToString();
+
+			if( attributeName.Equals( usingAlias, StringComparison.Ordinal ) ) {
+				return true;
+			}
+
+			const string attributeSuffix = "Attribute";
+			if( !usingAlias.EndsWith( attributeSuffix, StringComparison.Ordinal ) ) {
+				return false;
+			}
+
+			int shortUsingAliasLength = usingAlias.Length - attributeSuffix.Length;
+			if( shortUsingAliasLength != attributeName.Length ) {
+				return false;
+			}
+
+			bool equivalent = string.CompareOrdinal( attributeName, 0, usingAlias, 0, shortUsingAliasLength ) == 0;
+			return equivalent;
+		}
+
+		private IEnumerable<IdentifierNameSyntax> GetUsingAliases( AttributeSyntax attribute ) {
+
+			IEnumerable<UsingDirectiveSyntax> rootUsingDirectives = attribute.Ancestors()
+				.OfType<CompilationUnitSyntax>()

```

Comment:
`attribute.SyntaxTree.GetCompilationUnitRoot()` should usually be faster I think

Reply:
(Mostly just field accesses)

Reply:
Switched